runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkBatchSideInputHandlerFactory.java

```diff
@@ -0,0 +1,131 @@
+ */
+class FlinkBatchSideInputHandlerFactory implements MultimapSideInputHandlerFactory {
+
+  // Map from side input id to global PCollection id.
+  private final Map<SideInputId, PCollectionNode> sideInputToCollection;
+  private final RuntimeContext runtimeContext;
+
+  /**
+   * Creates a new state handler for the given stage. Note that this requires a traversal of the
+   * stage itself, so this should only be called once per stage rather than once per bundle.
+   */
+  static FlinkBatchSideInputHandlerFactory forStage(
+      ExecutableStage stage, RuntimeContext runtimeContext) {
+    ImmutableMap.Builder<SideInputId, PCollectionNode> sideInputBuilder = ImmutableMap.builder();
+    for (SideInputReference sideInput : stage.getSideInputs()) {
+      sideInputBuilder.put(
+          SideInputId.newBuilder()
+              .setTransformId(sideInput.transform().getId())
+              .setLocalName(sideInput.localName())
+              .build(),
+          sideInput.collection());
+    }
+    return new FlinkBatchSideInputHandlerFactory(sideInputBuilder.build(), runtimeContext);
+  }
+
+  private FlinkBatchSideInputHandlerFactory(
+      Map<SideInputId, PCollectionNode> sideInputToCollection, RuntimeContext runtimeContext) {
+    this.sideInputToCollection = sideInputToCollection;
+    this.runtimeContext = runtimeContext;
+  }
+
+  @Override
+  public <K, V, W extends BoundedWindow> MultimapSideInputHandler<K, V, W> forSideInput(
+      String transformId,
+      String sideInputId,
+      Coder<K> keyCoder,
+      Coder<V> valueCoder,
+      Coder<W> windowCoder) {
+    PCollectionNode collectionNode =
+        sideInputToCollection.get(
+            SideInputId.newBuilder().setTransformId(transformId).setLocalName(sideInputId).build());
+    checkArgument(collectionNode != null, "No side input for %s/%s", transformId, sideInputId);
+    List<Object> broadcastVariable = runtimeContext.getBroadcastVariable(collectionNode.getId());
+
+    ImmutableMultimap.Builder<SideInputKey<K, W>, V> multimap = ImmutableMultimap.builder();
+    for (Object element : broadcastVariable) {
+      WindowedValue<KV<K, V>> windowedValue = (WindowedValue<KV<K, V>>) element;
+      K key = windowedValue.getValue().getKey();
+      V value = windowedValue.getValue().getValue();
+
+      for (BoundedWindow boundedWindow : windowedValue.getWindows()) {
+        @SuppressWarnings("unchecked")
+        W window = (W) boundedWindow;
+        multimap.put(SideInputKey.of(key, window), value);
+      }
+    }
+
+    return new SideInputHandler<>(multimap.build());
+  }
+
+  private static class SideInputHandler<K, V, W extends BoundedWindow>
+      implements MultimapSideInputHandler<K, V, W> {
+
+    private final Multimap<SideInputKey<K, W>, V> collection;
+
+    private SideInputHandler(Multimap<SideInputKey<K, W>, V> collection) {
+      this.collection = collection;
+    }
+
+    @Override
+    public Iterable<V> get(K key, W window) {
```

Comment:
Just to clarify, is this the main input window or the side input window? (if it's the latter then it's fine)

Reply:
It's the side input window.